pkg/proxy/handlers/spacelister.go

```diff
@@ -36,25 +37,92 @@ func NewSpaceLister(app application.Application) *SpaceLister {
 }
 
 func (s *SpaceLister) HandleSpaceListRequest(ctx echo.Context) error {
+	userID, _ := ctx.Get(context.SubKey).(string)
+	username, _ := ctx.Get(context.UsernameKey).(string)
+
+	userSignup, err := s.GetSignupFunc(nil, userID, username, false)
+	if err != nil {
+		ctx.Logger().Error(errs.Wrap(err, "error retrieving userSignup"))
+		return errorResponse(ctx, apierrors.NewInternalError(err))
+	}
+	if userSignup == nil || userSignup.CompliantUsername == "" {
+		// account exists but the compliant username is not set yet, meaning it has not been fully provisioned yet
+		// return empty workspace list
+		return listWorkspaceResponse(ctx, []toolchainv1alpha1.Workspace{})
+	}
 
+	// list all user workspaces
 	workspaces, err := s.ListUserWorkspaces(ctx)
 	if err != nil {
 		return errorResponse(ctx, apierrors.NewInternalError(err))
 	}
+	return listWorkspaceResponse(ctx, workspaces)
+}
 
+func (s *SpaceLister) HandleSpaceGetRequest(ctx echo.Context) error {
+	userID, _ := ctx.Get(context.SubKey).(string)
+	username, _ := ctx.Get(context.UsernameKey).(string)
 	workspaceName := ctx.Param("workspace")
-	doGetWorkspace := len(workspaceName) > 0
 
-	if doGetWorkspace { // specific workspace requested
-		if len(workspaces) != 1 {
-			// not found
-			r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
-			return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
-		}
-		return getWorkspaceResponse(ctx, workspaces[0])
+	userSignup, err := s.GetSignupFunc(nil, userID, username, false)
+	if err != nil {
+		ctx.Logger().Error(errs.Wrap(err, "error retrieving userSignup"))
+		return errorResponse(ctx, apierrors.NewInternalError(err))
 	}
+	if userSignup == nil || userSignup.CompliantUsername == "" {
+		// account exists but the compliant username is not set yet, meaning it has not been fully provisioned yet
+		// return not found response when specific workspace request was issued
+		r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
+		return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
 
-	return listWorkspaceResponse(ctx, workspaces)
+	}
+
+	// get specific workspace
+	workspace, err := s.GetUserWorkspace(ctx, userSignup)
+	if err != nil {
+		return errorResponse(ctx, apierrors.NewInternalError(err))
+	}
+	if workspace == nil {
+		// not found
+		r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
+		return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
+	}
+	return getWorkspaceResponse(ctx, workspace)
+}
+
+func (s *SpaceLister) GetUserWorkspace(ctx echo.Context, signup *signup.Signup) (*toolchainv1alpha1.Workspace, error) {
+	murName := signup.CompliantUsername
+	spaceBinding, err := s.listSpaceBindingForUserAndSpace(ctx, murName)
```

Comment:
we should keep in mind the Space inheritance - if a user tries to get information about sub-space, then no SpaceBinding will be found. The logic needs to go up in the tree of the parents until it finds at least one.
It's fine to do it in the following PR :+1:

Reply:
Yeah, good point. Let me do that in the next PR. Thanks!